### PR TITLE
Add more binary ops implicit conversion

### PR DESCRIPTION
**Pull request by prometheansacrifice**
Handles more implicit abstract conversions during binary ops. Addresses #1109

### CODE REVIEW

**Comment by prometheansacrifice** [comment]
Removed WIP so that I can get some feedback on this PR.

When running the error handler tests, amidst the logs I saw this,

```
test/error-handler/call2.js                        
test/error-handler/class.js                        
Unexpected error: serialized 7 of 8                
test/error-handler/conditional-return.js           
test/error-handler/forInStatement.js               
test/error-handler/forOfStatement.js 
```

But the results of the test said 36/36 100%. Is this why the CI is failing?

**Comment by hermanventer** [comment]
The error handler tests should fail, but this is probably unrelated to your changes. The other test failure is in test262 and need attention.

File: src/domains/ValuesDomain.js [hunk 1]
```diff
@@ -145,6 +145,13 @@ export default class ValuesDomain {
       return Add(realm, lnum, rnum);
     } else if (op === "<" || op === ">" || op === ">=" || op === "<=") {
       // ECMA262 12.10.3
+      let lprim = To.ToPrimitiveOrAbstract(realm, lval);
```
Comment (hermanventer) [inline src/domains/ValuesDomain.js]: Doing this here and then calling AbstractRelationalComparison below can cause side effects to happen more than once if lval has a side-effecting implicit conversion. This is probably why there are test262 failures.

**Comment by hermanventer** [comment]
Any intention to come back to this?

**Comment by prometheansacrifice** [comment]
I do. I want to take another shot this weekend. Hope it's okay.

**Comment by prometheansacrifice** [comment]
@hermanventer I had trouble with `==` and `!=`: `AbstractEqualityComparison` returns a boolean, but I needed to return an `AbstractValue`

L245 in ./methods/abstract.js,
Idea was to change
```
  if ((x instanceof StringValue || x instanceof NumberValue || x instanceof SymbolValue) && y instanceof ObjectValue) {
    return AbstractEqualityComparison(realm, x, To.ToPrimitive(realm, y));
  }
```
to
```
  if ((x instanceof StringValue || x instanceof NumberValue || x instanceof SymbolValue) && y instanceof ObjectValue) {
    const py = To.ToPrimitiveOrAbstract(realm, y);
    if (py instanceof AbstractValue) {
       return AbstractValue.createFromBinaryOp(realm, op, x, py);
    }
    return AbstractEqualityComparison(realm, x, py);
  }
```

File: src/methods/abstract.js [hunk 2]
```diff
@@ -107,24 +107,34 @@ export function AbstractRelationalComparison(
...
-    py = To.ToPrimitive(realm, y, "number");
+    py = To.ToPrimitiveOrAbstract(realm, y, "number");
 
     // b. Let px be ? ToPrimitive(x, hint Number).
-    px = To.ToPrimitive(realm, x, "number");
+    px = To.ToPrimitiveOrAbstract(realm, x, "number");
+  }
+
+  if (px instanceof AbstractValue || py instanceof AbstractValue) {
+    invariant(op !== undefined, "operator must be provided if abstract values are to be compared");
```
Comment (hermanventer) [inline src/methods/abstract.js]: It seems that caller's will have a hard time ensuring that this invariant is never violated, unless they always pass in op. So, shouldn't op be non optional?

**Comment by hermanventer** [comment]
It seems that the only places that call AbstractEqualityComparison would be happy to get a Value rather than a boolean.

File: src/methods/abstract.js (hunk 2 above)
Comment (prometheansacrifice) [reply src/methods/abstract.js]: I was not sure of touching `./intrinsics/ecma262/ArrayPrototype.js`

https://github.com/facebook/prepack/blob/master/src/intrinsics/ecma262/ArrayPrototype.js#L1353-L1354

https://github.com/facebook/prepack/blob/master/src/intrinsics/ecma262/ArrayPrototype.js#L1357-L1358

Is it okay update the call over there?

**Comment by prometheansacrifice** [comment]
> It seems that the only places that call AbstractEqualityComparison would be happy to get a Value rather than a boolean.

Making the change an updating the PR.

File: src/methods/abstract.js (hunk 2 above)
Comment (hermanventer) [reply src/methods/abstract.js]: Yes. The operator use would be "<". The call sites should also assert that their return values are BooleanValue instances.

**Comment by hermanventer** [comment]
The PR seems not to have updated.

**Comment by prometheansacrifice** [comment]
@hermanventer It hasn't. I'm sorry I should have marked it WIP.

**Comment by prometheansacrifice** [comment]
@hermanventer Added support for `!=` and `==`. Removed the invariant and made `op` param mandatory too.

File: test/serializer/abstract/ImplicitAbstractConversion5.js [hunk 3]
```diff
@@ -0,0 +1,4 @@
+let x = global.__abstract ? __abstract("number", "(3)") : 3;
+let ob =  { valueOf: () => x };
+let y = 2 ** ob;
```
Comment (hermanventer) [inline test/serializer/abstract/ImplicitAbstractConversion5.js]: Add a test with an abstract base.

File: test/serializer/abstract/ImplicitAbstractConversion4.js [hunk 4]
```diff
@@ -0,0 +1,4 @@
+let x = global.__abstract ? __abstract("number", "(3)") : 3;
+let ob =  { valueOf: () => x };
+let y = 123 > ob;
```
Comment (hermanventer) [inline test/serializer/abstract/ImplicitAbstractConversion4.js]: Add a test with <

**Comment by prometheansacrifice** [comment]
@hermanventer Can you please take a look at the PR again? Sorry for disappearing for so long. I think the coverage has increased 2% in the ValuesDomain file.

**Pull request closed without merging**